### PR TITLE
Subtract items in other customers' cart from stock

### PR DESCRIPTION
**Pull request by sikachu**
First part of reservation feature.

This is done by counting all items that are orders that are incomplete and subtract it from in-hand count. Note that this already take current customer's cart into account, meaning that their order should process successfully even though it blocks other users from making a purchase.

### CODE REVIEW

File: app/models/spree/line_item_decorator.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module Spree
+  LineItem.class_eval do
+    # Override
```
Comment (elle) [inline app/models/spree/line_item_decorator.rb]: Remove comment?

File: app/models/spree/stock/availability_validator_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+module Spree
+  module Stock
+    AvailabilityValidator.class_eval do
+      # Override
```
Comment (elle) [inline app/models/spree/stock/availability_validator_decorator.rb]: Remove comment?

File: app/models/spree/line_item_decorator.rb (hunk 1 above)
Comment (sikachu) [inline app/models/spree/line_item_decorator.rb]: Sure. I feel like they are there just so you could see that this might break stuff ... but I guess that's not necessary. Removing them.

File: spec/features/customer_checkouts_spec.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    visit spree.root_path
+    click_on product.name
+    click_button "Add To Cart"
+    fill_in line_item_quantity_field, with: "2"
+    click_button "Update"
+
+    expect(page).to have_content("not available")
+  end
+
```
Comment (elle) [inline spec/features/customer_checkouts_spec.rb]: Should spec helper methods be private?

File: spec/features/customer_checkouts_spec.rb [hunk 4]
```diff
@@ -0,0 +1,89 @@
+require "spec_helper"
+
+RSpec.feature "Customer checkouts" do
+  let!(:product) { create(:product_in_stock) }
+  let!(:shipping_method) { create(:shipping_method) }
+  let!(:payment_method) { create(:check_payment_method) }
+  let!(:store) { create(:store) }
+
+  before do
+    shipping_method.calculator.update!(preferred_amount: 10)
```
Comment (elle) [inline spec/features/customer_checkouts_spec.rb]: Could we stub this instead of calling update to the db again?

File: spec/features/customer_checkouts_spec.rb [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    product.shipping_category = shipping_method.shipping_categories.first
+    product.master.stock_items.first.update!(backorderable: false)
+    product.save!
```
Comment (elle) [inline spec/features/customer_checkouts_spec.rb]: Is there any way we can set up this product once, using the factory, without having to save it again?

File: spec/models/spree/line_item_spec.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+      order = build_stubbed(:order)
+      quantifier = instance_double(
+        Spree::Stock::QuantifierWithPending,
+        can_supply?: true
+      )
+      allow(Spree::Stock::QuantifierWithPending).
+        to receive(:new).and_return(quantifier)
+      line_item = build(:line_item, variant: variant, order: order, quantity: 1)
+
+      line_item.sufficient_stock?
```
Comment (elle) [inline spec/models/spree/line_item_spec.rb]: Do we want to add an expectation to this method call?

File: spec/models/spree/line_item_spec.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
+require "spec_helper"
+
+RSpec.describe Spree::LineItem do
+  describe "#sufficient_stock?" do
```
Comment (elle) [inline spec/models/spree/line_item_spec.rb]: Can we also test the negative outcome?

File: spec/models/spree/stock/availability_validator_spec.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+      order = build_stubbed(:order)
+      quantifier = instance_double(
+        Spree::Stock::QuantifierWithPending,
+        can_supply?: true
+      )
+      allow(Spree::Stock::QuantifierWithPending).
+        to receive(:new).and_return(quantifier)
+      line_item = build(:line_item, variant: variant, order: order)
+
+      Spree::Stock::AvailabilityValidator.new.item_available?(line_item, 1)
```
Comment (elle) [inline spec/models/spree/stock/availability_validator_spec.rb]: Pretty much same comments: do we want to expect this to return true?
And do we want to test the negative option?

File: spec/models/spree/stock/quantifier_with_pending_spec.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  describe "#total_on_hand" do
+    context "without giving a pending order" do
+      it "returns remaining stock including reservations in customers' cart" do
+        create_pending_order(quantity: 1)
+        create_pending_order(quantity: 2)
+        quantifier = Spree::Stock::QuantifierWithPending.new(variant)
+
+        expect(quantifier.total_on_hand).to eq 7
```
Comment (elle) [inline spec/models/spree/stock/quantifier_with_pending_spec.rb]: Do we want to check that the `total_on_hand` can never be negative?

File: spec/features/customer_checkouts_spec.rb (hunk 4 above)
Comment (sikachu) [inline spec/features/customer_checkouts_spec.rb]: Nope. It's a feature spec, so it has to hit the live database.

File: spec/features/customer_checkouts_spec.rb (hunk 5 above)
Comment (sikachu) [inline spec/features/customer_checkouts_spec.rb]: I don't think so. I actually copied this from [spree test suite](https://github.com/spree/spree/blob/d99fa9bc33b95affb77dd085ff24c06e420e586a/frontend/spec/features/checkout_spec.rb#L43) and this is here just to prevent regression.

File: spec/features/customer_checkouts_spec.rb (hunk 3 above)
Comment (sikachu) [inline spec/features/customer_checkouts_spec.rb]: I don't think they have to be. I look at some code (in Upcase, for example) and I don't see a `private` marker in there either.

I don't mind adding it though, but it seems unnecessary.

File: spec/models/spree/line_item_spec.rb (hunk 7 above)
Comment (sikachu) [inline spec/models/spree/line_item_spec.rb]: This method calls `QuantifierWithPending#can_supply?` internally, so I initially thought we don't need it.

I think what I'm going to do is to adding another expectation to make sure that the method return what's returned from `can_supply?`.

File: spec/models/spree/stock/quantifier_with_pending_spec.rb (hunk 9 above)
Comment (sikachu) [inline spec/models/spree/stock/quantifier_with_pending_spec.rb]: Good point. In some case with a race condition, it could be negative ... Let me adding another test case to cover that.

Comment (sikachu) [inline spec/models/spree/stock/quantifier_with_pending_spec.rb]: This is done below.

**Comment by sikachu** [comment]
@elle would you mind giving this another look when you have a chance?

File: app/models/spree/stock/quantifier_with_pending.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
+module Spree
+  module Stock
+    class QuantifierWithPending < Quantifier
+      def initialize(variant, current_order = nil)
+        super(variant)
+        @current_order = current_order
+      end
+
+      def total_on_hand
+        [super - total_pending, 0].max
```
Comment (elle) [inline app/models/spree/stock/quantifier_with_pending.rb]: :+1:

File: spec/models/spree/line_item_spec.rb [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    it "uses Stock::QuantifierWithPending" do
+      variant = build_stubbed(:variant)
+      order = build_stubbed(:order)
+      can_supply_result = Object.new
```
Comment (elle) [inline spec/models/spree/line_item_spec.rb]: Just a question: why is `Object.new` better than `true` or `false`?

**Comment by elle** [comment]
Except to one question, LGTM

File: spec/models/spree/line_item_spec.rb (hunk 11 above)
Comment (sikachu) [inline spec/models/spree/line_item_spec.rb]: It's because with `true` or `false` you can actually change the method to do something like:

```ruby
def sufficient_stock?
  quantifier.can_supply?(1)
  false
end
```

and your test might turn green. With stubbing method and returning an arbitrary object (aka. poison object) we can confirm that the method actually return what it has been given from another method.
